src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+            if hours > 999999:
+                self.renderable = "+999999h"
+            elif hours > 99:
+                self.renderable = f"{hours}h"
+            else:
+                self.renderable = f"{hours:02}:{minutes:02}:{seconds:02}"
+        self.refresh()
+
+
+class ProgressBar(Widget, can_focus=False):
+    """A progress bar widget."""
+
+    DEFAULT_CSS = """
+    Horizontal {
+        width: auto;
+        height: auto;
+    }
+    ProgressBar {
+        width: auto;
+        height: 1;
+    }
+    """
+
+    progress: reactive[float] = reactive(0.0)
+    """The progress so far, in number of steps."""
+    total: reactive[float | None] = reactive[Optional[float]](None)
+    """The total number of steps associated with this progress bar, when known.
+
+    The value `None` will render an indeterminate progress bar.
+    Once `total` is set to a numerical value, it cannot be set back to `None`.
+    """
+    _percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed.
+
+    We keep this private to prevent the user from having multiple reactive attributes
+    that can be used to advance the progress of the bar.
+    """
+    _started = False
+    """Whether the total was set to a non-`None` value before."""
+
+    class Completed(Message):
+        """Sent when the progress reaches 100%.
+
+        Can be handled using `on_progress_bar_completed` in a subclass of
+        [`ProgressBar`][textual.widgets.ProgressBar] or in a parent widget in the DOM.
+        """
+
+        def __init__(self, progress_bar: ProgressBar) -> None:
+            """Create the message instance.
+
+            Args:
+                progress_bar: The progress bar that reached 100% completion.
+            """
+            self.progress_bar = progress_bar
+            super().__init__()
+
+    class Started(Message):
+        """Sent when the progress bar starts tracking progress.
+
+        This message is sent when the total progress is set to a non-`None` value.
+        This message may be sent _before_ the progress bar is mounted if the progress
+        bar is instantiated with a non-`None` value for the total number of steps.
+
+        Can be handled using `on_progress_bar_started` in a subclass of
+        [`ProgressBar`][textual.widgets.ProgressBar] or in a parent widget in the DOM.
+        """
+
+        def __init__(self, progress_bar: ProgressBar, total: float) -> None:
+            """Create a message instance.
+
+            Args:
+                progress_bar: The progress bar that started tracking progress.
+                total: The total number of steps that the bar will track.
+            """
+            self.progress_bar = progress_bar
+            self.total = total
+            super().__init__()
+
+    def __init__(
+        self,
+        total: float | None = None,
+        *,
+        hide_bar: bool = False,
```

Comment:
I think these parameter may be better as `show_*`, otherwise it feels like a double negetive.